### PR TITLE
Bound Initial Sync Cache Size

### LINKED ISSUE #4813: Initial Sync Cache State Leads to OOMs
Using the ```initial-sync-cache-state``` flags leads to  OOMs when syncing blocks in a period without finality. This is as we cache the relevant fields of each state in a slot. Without finality there is no pruning of states, so all these fields are kept on the heap and leads to the heap size quickly ballooning.

A solution for this would be to handle periods without finality properly and not keep all states in memory.

### PR DESCRIPTION
**Pull request by nisdas**
Resolves #4813
- [x] Only Cache Boundary States instead of all unfinalized states.
- [x] Enforce Cache Limit so that states are flushed from memory to disk, in the
event caches are full.
- [x] Filter out boundary candidate states at the start of each epoch. 
- [x] Use memory pool for Randao mixes field.
- [x] Add flag to toggle with garbage collector.

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -243,6 +247,26 @@ func (s *Service) onBlockInitialSyncStateTransition(ctx context.Context, signed
 		return errors.Wrap(err, "could not save finalized checkpoint")
 	}
 
+	if featureconfig.Get().InitSyncCacheState {
+		numOfStates := len(s.initSyncState)
+		if numOfStates > initialSyncCacheSize {
+			stateSlice := make([][32]byte, 0, numOfStates)
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go]: Can you please refactor this into its own function?

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -243,6 +247,26 @@ func (s *Service) onBlockInitialSyncStateTransition(ctx context.Context, signed
...
+			for rt := range s.initSyncState {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go]: Tests?

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+		delete(s.initSyncState, rt)
+	}
+	s.boundaryRoots = s.boundaryRoots[numOfStates-minimumCacheSize:]
+	return nil
+}
+
+// filter out boundary candidates from our currently processed batch of states.
+func (s *Service) filterBoundaryCandidates(ctx context.Context, root [32]byte, postState *stateTrie.BeaconState) {
+	stateSlice := make([][32]byte, 0, len(s.initSyncState))
+	// add slots to map and add epoch boundary states
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
	// Add slots to the map and add epoch boundary states to the slice.
```

File: beacon-chain/blockchain/process_block.go [hunk 4]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+	// to the slice
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
```

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+	for rt := range s.initSyncState {
+		stateSlice = append(stateSlice, rt)
+	}
+
+	sort.Slice(stateSlice, func(i int, j int) bool {
+		return s.initSyncState[stateSlice[i]].Slot() < s.initSyncState[stateSlice[j]].Slot()
+	})
+	epochLength := params.BeaconConfig().SlotsPerEpoch
+
+	// only trigger on epoch start
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
	// Only trigger on epoch start.
```

File: beacon-chain/blockchain/process_block.go [hunk 6]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+	if helpers.IsEpochStart(postState.Slot()) {
+		if len(s.boundaryRoots) > 0 {
+			// retrieve previous boundary root
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
			// Retrieve previous boundary root.
```

File: beacon-chain/blockchain/process_block.go [hunk 7]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+			previousBoundaryRoot := s.boundaryRoots[len(s.boundaryRoots)-1]
+			previousSlot := s.initSyncState[previousBoundaryRoot].Slot()
+
+			// round up slot number to account for skipped slots
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
			// Round up slot number to account for skipped slots.
```

File: beacon-chain/blockchain/process_block.go [hunk 8]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+			previousSlot = helpers.RoundUpToNearestEpoch(previousSlot)
+			if postState.Slot()-previousSlot > epochLength {
+				targetSlot := postState.Slot()
+				tempRoots := [][32]byte{}
+
+				// loop through current states to filter for valid boundary states
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
				// Loop through current states to filter for valid boundary states.
```

File: beacon-chain/blockchain/process_block.go [hunk 9]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+				for i := len(stateSlice) - 1; stateSlice[i] != previousBoundaryRoot && i >= 0; i-- {
+					currentSlot := s.initSyncState[stateSlice[i]].Slot()
+					// store states from the start/end of epochs
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
					// Store states from the start/end of epochs.
```

File: beacon-chain/blockchain/process_block.go [hunk 10]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+					if currentSlot-1 == targetSlot-epochLength ||
+						currentSlot+1 == targetSlot-epochLength {
+						tempRoots = append(tempRoots, stateSlice[i])
+						continue
+					}
+					// skip if the current slot is larger than the previous epoch
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
					// Skip if the current slot is larger than the previous epoch
```

File: beacon-chain/blockchain/process_block.go [hunk 11]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+					// boundary
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
					// boundary.
```

File: beacon-chain/blockchain/process_block.go [hunk 12]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+					if currentSlot > targetSlot-epochLength {
+						continue
+					}
+					tempRoots = append(tempRoots, stateSlice[i])
+
+					// switch target slot if the current slot is more than
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
					// Switch target slot if the current slot is greater than
```

File: beacon-chain/blockchain/process_block.go [hunk 13]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+					// 1 epoch boundary from the previously saved boundary slot
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
					// 1 epoch boundary from the previously saved boundary slot.
```

File: beacon-chain/blockchain/process_block.go [hunk 14]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+					if currentSlot > previousSlot+epochLength {
+						currentSlot = helpers.RoundUpToNearestEpoch(currentSlot)
+						targetSlot = currentSlot
+						continue
+					}
+					break
+				}
+				// reverse to append the roots in ascending order corresponding
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
				// Reverse to append the roots in ascending order corresponding
```

File: beacon-chain/db/kv/state.go [hunk 15]
```diff
@@ -120,6 +120,34 @@ func (k *Store) SaveState(ctx context.Context, state *state.BeaconState, blockRo
 	})
 }
 
+// SaveStates stores multiple states to the db using block's signing root which was used to generate the state.
```
Comment (0xKiwi) [inline beacon-chain/db/kv/state.go]:
```suggestion
// SaveStates stores multiple states to the db using the block's signing root which was used to generate the state.
```

File: beacon-chain/main.go [hunk 16]
```diff
@@ -126,6 +127,8 @@ func main() {
 			}
 		}
 
+		runtimeDebug.SetGCPercent(ctx.GlobalInt(flags.SetGCPercent.Name))
+
```
Comment (0xKiwi) [inline beacon-chain/main.go]:
```suggestion
```

File: beacon-chain/state/getters.go [hunk 17]
```diff
@@ -2,8 +2,11 @@ package state
 
 import (
 	"fmt"
+
 	"github.com/prysmaticlabs/go-bitfield"
 
+	"github.com/prysmaticlabs/prysm/shared/memorypool"
+
```
Comment (0xKiwi) [inline beacon-chain/state/getters.go]:
```suggestion
```

File: beacon-chain/state/types.go [hunk 18]
```diff
@@ -4,6 +4,8 @@ import (
 	"runtime"
 	"sync"
 
+	"github.com/prysmaticlabs/prysm/shared/memorypool"
+
```
Comment (0xKiwi) [inline beacon-chain/state/types.go]:
```suggestion
```

File: beacon-chain/blockchain/process_block.go [hunk 19]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
 
 	return nil
 }
+
+func (s *Service) persistCachedStates(ctx context.Context, numOfStates int) error {
+	log.Error("persist cache triggered")
+	oldStates := make([]*stateTrie.BeaconState, 0, numOfStates)
+
+	// add slots to map and add epoch boundary states
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
	// Add slots to the map and add epoch boundary states to the slice.
```

File: beacon-chain/blockchain/process_block.go [hunk 20]
```diff
@@ -298,3 +313,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+	// to the slice
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]:
```suggestion
```

File: beacon-chain/blockchain/process_block.go [hunk 21]
```diff
@@ -298,3 +313,113 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
 
 	return nil
 }
+
+func (s *Service) persistCachedStates(ctx context.Context, numOfStates int) error {
+	log.Error("persist cache triggered")
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]: Why is this log an error? Maybe it'll confuse users.

File: beacon-chain/blockchain/process_block.go [hunk 22]
```diff
@@ -298,3 +313,113 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+		stateSlice = append(stateSlice, rt)
+	}
+
+	sort.Slice(stateSlice, func(i int, j int) bool {
+		return s.initSyncState[stateSlice[i]].Slot() < s.initSyncState[stateSlice[j]].Slot()
+	})
+	epochLength := params.BeaconConfig().SlotsPerEpoch
+
+	// Only trigger on epoch start.
+	if helpers.IsEpochStart(postState.Slot()) {
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block.go]: Can you inverse this and make the function return here so there isn't as many nested if statements?

File: beacon-chain/blockchain/process_block.go [hunk 23]
```diff
@@ -298,3 +314,115 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
 
 	return nil
 }
+
+func (s *Service) persistCachedStates(ctx context.Context, numOfStates int) error {
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go]: it doesnt prune them, we save it to disk. If you look below we call `SaveStates`

File: beacon-chain/blockchain/process_block_test.go [hunk 24]
```diff
@@ -537,6 +532,47 @@ func TestFilterBlockRoots_CanFilter(t *testing.T) {
...
+			t.Errorf("State with root of %#x , could not be retrieved: %v", root, err)
+		}
+		if state == nil {
+			t.Errorf("State with root of %#x , does not exist", root)
+		}
+		if state.Slot() != i {
+			t.Errorf("Incorrect slot retrieved. Wanted %d but got %d", i, state.Slot())
+		}
+	}
+
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block_test.go]:
```suggestion
```

File: beacon-chain/blockchain/process_block_test.go [hunk 25]
```diff
@@ -650,6 +686,194 @@ func TestFillForkChoiceMissingBlocks_FilterFinalized(t *testing.T) {
...
+		}
+	}
+	lastIndex := len(service.boundaryRoots) - 1
+	for i := uint64(500); i < 2000; i++ {
+		st.SetSlot(i)
+		root := [32]byte{}
+		copy(root[:], bytesutil.Bytes32(i))
+		service.initSyncState[root] = st.Copy()
+	}
+	// set current state
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block_test.go]:
```suggestion
	// Set current state.
```

File: beacon-chain/blockchain/process_block_test.go [hunk 26]
```diff
@@ -650,6 +686,194 @@ func TestFillForkChoiceMissingBlocks_FilterFinalized(t *testing.T) {
...
+	for i := uint64(500); i < 2000; i++ {
+		st.SetSlot(i)
+		root := [32]byte{}
+		copy(root[:], bytesutil.Bytes32(i))
+		// save only for offsetted slots
+		if helpers.IsEpochStart(i + 10) {
+			service.initSyncState[root] = st.Copy()
+		}
+	}
+	// set current state
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block_test.go]:
```suggestion
	// Set current state.
```

File: beacon-chain/blockchain/process_block_test.go [hunk 27]
```diff
@@ -650,6 +686,194 @@ func TestFillForkChoiceMissingBlocks_FilterFinalized(t *testing.T) {
...
+		t.Fatal("Wanted non zero added boundary roots")
+	}
+	for _, rt := range service.boundaryRoots[lastIndex+1:] {
+		st, ok := service.initSyncState[rt]
+		if !ok {
+			t.Error("Root doen't exist in cache map")
+			continue
+		}
+		if st.Slot() >= 500 {
+			// ignore head boundary root
```
Comment (0xKiwi) [inline beacon-chain/blockchain/process_block_test.go]:
```suggestion
			// Ignore head boundary root.
```

File: beacon-chain/state/getters.go [hunk 28]
```diff
@@ -2,8 +2,11 @@ package state
 
 import (
 	"fmt"
+
 	"github.com/prysmaticlabs/go-bitfield"
```
Comment (0xKiwi) [inline beacon-chain/state/getters.go]:
```suggestion
```

File: beacon-chain/db/kv/state.go [hunk 29]
```diff
@@ -120,6 +120,34 @@ func (k *Store) SaveState(ctx context.Context, state *state.BeaconState, blockRo
 	})
 }
 
+// SaveStates stores multiple states to the db using the block's signing root which was used to generate the state.
```
Comment (nisdas) [inline beacon-chain/db/kv/state.go]:
```suggestion
// SaveStates stores multiple states to the db using the provided corresponding roots.
```

File: beacon-chain/blockchain/process_block.go [hunk 30]
```diff
@@ -298,3 +313,113 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
...
+		stateSlice = append(stateSlice, rt)
+	}
+
+	sort.Slice(stateSlice, func(i int, j int) bool {
+		return s.initSyncState[stateSlice[i]].Slot() < s.initSyncState[stateSlice[j]].Slot()
+	})
+	epochLength := params.BeaconConfig().SlotsPerEpoch
+
+	// Only trigger on epoch start.
+	if !helpers.IsEpochStart(postState.Slot()) {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go]: why not do it at the start of the function?

File: beacon-chain/blockchain/process_block.go [hunk 31]
```diff
@@ -298,3 +313,113 @@ func (s *Service) insertBlockToForkChoiceStore(ctx context.Context, blk *ethpb.B
 
 	return nil
 }
+
+func (s *Service) persistCachedStates(ctx context.Context, numOfStates int) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go]: function comment please, and maybe a better place for this and below are in `process_block_helpers.go` ?

File: beacon-chain/core/helpers/slot_epoch.go [hunk 32]
```diff
@@ -106,3 +106,12 @@ func VerifySlotTime(genesisTime uint64, slot uint64) error {
 func SlotsSince(time time.Time) uint64 {
 	return uint64(roughtime.Since(time).Seconds()) / params.BeaconConfig().SecondsPerSlot
 }
+
+// RoundUpToNearestEpoch rounds up the provided slot value to the nearest epoch
```
Comment (terencechain) [inline beacon-chain/core/helpers/slot_epoch.go]:
```suggestion
// RoundUpToNearestEpoch rounds up the provided slot value to the nearest epoch.
```

File: beacon-chain/core/helpers/slot_epoch.go [hunk 33]
```diff
@@ -106,3 +106,12 @@ func VerifySlotTime(genesisTime uint64, slot uint64) error {
...
+func RoundUpToNearestEpoch(slot uint64) uint64 {
```
Comment (terencechain) [inline beacon-chain/core/helpers/slot_epoch.go]: Test for this?

File: beacon-chain/blockchain/process_block.go (hunk 31 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go]: or even better `init_sync_process_block.go` or something more initial sync related. If this is only used by initial sync

File: beacon-chain/core/helpers/slot_epoch.go (hunk 33 above)
Comment (nisdas) [reply beacon-chain/core/helpers/slot_epoch.go]: done , added in

File: beacon-chain/blockchain/process_block.go (hunk 31 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go]: moved all functions to a separate file

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 34]
```diff
@@ -0,0 +1,178 @@
...
+			// Reverse to append the roots in ascending order corresponding
+			// to the respective slots.
+			tempRoots = bytesutil.ReverseBytes32Slice(tempRoots)
+			s.boundaryRoots = append(s.boundaryRoots, tempRoots...)
+		}
+	}
+	s.boundaryRoots = append(s.boundaryRoots, root)
+	s.pruneOldStates()
+	s.pruneNonBoundaryStates()
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/init_sync_process_block.go]:
```suggestion
```

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 35]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+// prune for states past the current finalized checkpoint.
+func (s *Service) pruneOldStates() {
+	prunedBoundaryRoots := [][32]byte{}
+	for _, rt := range s.boundaryRoots {
+		st, ok := s.initSyncState[rt]
+		// skip non-existent roots
```
Comment (rauljordan) [inline beacon-chain/blockchain/init_sync_process_block.go]:
```suggestion
		// Skip non-existent roots.
```

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 36]
```diff
@@ -0,0 +1,177 @@
+package blockchain
+
+import (
+	"context"
+	"sort"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/stategen"
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/init_sync_process_block.go]:
```suggestion
```

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 37]
```diff
@@ -0,0 +1,177 @@
...
+const maxCacheSize = 70
+const initialSyncCacheSize = 45
+const minimumCacheSize = initialSyncCacheSize / 3
+
+func (s *Service) persistCachedStates(ctx context.Context, numOfStates int) error {
+	oldStates := make([]*stateTrie.BeaconState, 0, numOfStates)
+
+	// Add slots to the map and add epoch boundary states to the slice.
+	for _, rt := range s.boundaryRoots[:numOfStates-minimumCacheSize] {
```
Comment (rauljordan) [inline beacon-chain/blockchain/init_sync_process_block.go]: Can this panic out of range?

File: beacon-chain/stategen/service.go [hunk 38]
```diff
@@ -1,6 +1,14 @@
 package stategen
 
 import (
+	"context"
+
+	"github.com/pkg/errors"
+
```
Comment (rauljordan) [inline beacon-chain/stategen/service.go]:
```suggestion
```

File: beacon-chain/stategen/service.go [hunk 39]
```diff
@@ -1,6 +1,14 @@
...
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/go-ssz"
+	stateTrie "github.com/prysmaticlabs/prysm/beacon-chain/state"
+
```
Comment (rauljordan) [inline beacon-chain/stategen/service.go]:
```suggestion
```

File: shared/memorypool/memorypool.go [hunk 40]
```diff
@@ -0,0 +1,27 @@
+package memorypool
```
Comment (rauljordan) [inline shared/memorypool/memorypool.go]: @nisdas this stuff seems unrelated to the PR, can it be made into a separate PR?

File: beacon-chain/stategen/service_test.go [hunk 41]
```diff
@@ -0,0 +1,74 @@
+package stategen
+
+import (
+	"context"
+	"testing"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"gopkg.in/d4l3k/messagediff.v1"
+
```
Comment (rauljordan) [inline beacon-chain/stategen/service_test.go]:
```suggestion
```

File: beacon-chain/stategen/service.go [hunk 42]
```diff
@@ -16,3 +24,21 @@ func New(db db.NoHeadAccessDatabase) *State {
 		beaconDB: db,
 	}
 }
+
+// GenerateState generates a state from the provided pre-state till the provided block.
+func (s *State) GenerateState(ctx context.Context, preState *stateTrie.BeaconState, endBlock *ethpb.SignedBeaconBlock) (*stateTrie.BeaconState, error) {
```
Comment (rauljordan) [inline beacon-chain/stategen/service.go]: Love how clear this function is, everything makes total sense

File: beacon-chain/flags/base.go [hunk 43]
```diff
@@ -74,6 +74,12 @@ var (
 		Usage: "The eth1 block in which the deposit contract was deployed.",
 		Value: 1960177,
 	}
+	// SetGCPercent is the percentage of current live allocations at which the garbage collector is to run.
+	SetGCPercent = cli.IntFlag{
```
Comment (rauljordan) [inline beacon-chain/flags/base.go]: Can this be a separate PR?

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 44]
```diff
@@ -0,0 +1,177 @@
...
+		previousBoundaryRoot := s.boundaryRoots[len(s.boundaryRoots)-1]
+		previousSlot := s.initSyncState[previousBoundaryRoot].Slot()
+
+		// Round up slot number to account for skipped slots.
+		previousSlot = helpers.RoundUpToNearestEpoch(previousSlot)
+		if postState.Slot()-previousSlot > epochLength {
+			targetSlot := postState.Slot()
+			tempRoots := [][32]byte{}
+
+			// Loop through current states to filter for valid boundary states.
```
Comment (rauljordan) [inline beacon-chain/blockchain/init_sync_process_block.go]: Can this be put into a helper func? the func is getting too big here

**Comment by rauljordan** [comment]
Lookin good @nisdas just had comments about simplification and moving extraneous features into smaller PRs. Overall this is a very non-trivial feature that can have unexpected consequences so we can run this in experimental for a while.

File: shared/memorypool/memorypool.go (hunk 40 above)
Comment (nisdas) [reply shared/memorypool/memorypool.go]: alright, done #4896

File: beacon-chain/blockchain/init_sync_process_block.go (hunk 37 above)
Comment (nisdas) [reply beacon-chain/blockchain/init_sync_process_block.go]: if used normally without context , yes. But we only use it when we have `len(s.boundaryRoots) > 45`, so this will never happen in how it is currently used.

File: beacon-chain/flags/base.go (hunk 43 above)
Comment (nisdas) [reply beacon-chain/flags/base.go]: done, #4897

File: beacon-chain/stategen/service.go [hunk 45]
```diff
@@ -16,3 +22,21 @@ func New(db db.NoHeadAccessDatabase) *State {
...
```
Comment (terencechain) [inline beacon-chain/stategen/service.go]: can you put this back in block chain `service.go`, you can combine it with: https://github.com/prysmaticlabs/prysm/pull/4844/files#diff-c2ffc98dd863a51dc11c4843367b82b6R160

this API pattern with `preState` and `endBlock` seem to be only serving block chain service specific use case, because of that, we shouldn't put it here.

I'll export `loadBlocks` and `replayBlocks` via an interface for you

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 46]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	endBlock, err := s.beaconDB.Block(ctx, endRoot)
+	if err != nil {
+		return nil, err
+	}
+	if endBlock == nil {
+		return nil, errors.New("provided block root does not have block saved in the db")
+	}
+	stGen := stategen.New(s.beaconDB)
```
Comment (terencechain) [inline beacon-chain/blockchain/init_sync_process_block.go]: can you initialize this properly in node.go and pass it down to the service struct so `stGen` is part of *Service?

File: beacon-chain/stategen/service.go (hunk 45 above)
Comment (nisdas) [reply beacon-chain/stategen/service.go]: Wouldn't something like this also be useful for other services, ex: Sync,etc ? You can generate a new state from a provided pre-state

Comment (terencechain) [reply beacon-chain/stategen/service.go]: I agree, but we don't know what the API is for that yet. I think until sync actually needs it / uses it. It's better to constraint this within the block chains service. 
Or else it may introduce disparity where chain service uses one API to generate state and sync service uses another API to generate state and both APIs reside in stategen service

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 47]
```diff
@@ -0,0 +1,191 @@
...
+		return s.initSyncState[stateSlice[i]].Slot() < s.initSyncState[stateSlice[j]].Slot()
+	})
+	epochLength := params.BeaconConfig().SlotsPerEpoch
+
+	if len(s.boundaryRoots) > 0 {
+		// Retrieve previous boundary root.
+		previousBoundaryRoot := s.boundaryRoots[len(s.boundaryRoots)-1]
+		previousState, ok := s.initSyncState[previousBoundaryRoot]
+		if !ok {
+			// remove the non-existent root and exit filtering.
```
Comment (terencechain) [inline beacon-chain/blockchain/init_sync_process_block.go]:
```suggestion
			// Remove the non-existent root and exit filtering.
```
